Solutions/Azure Active Directory/Analytic Rules/MaliciousOAuthApp_O365AttackToolkit.yaml

```diff
@@ -24,18 +24,20 @@ relevantTechniques:
 query: |
   let detectionTime = 1d;
   let joinLookback = 14d;
+  let o365_attack_v1 = dynamic(["contacts.read","user.read","mail.read","notes.read.all","mailboxsettings.readwrite","Files.ReadWrite.All"]);
+  let o365_attack_v2 = dynamic(["contacts.read","user.read","mail.read","mail.send","Files.ReadWrite.All","files.read","files.read.all"]);
```

Comment:
It would be great if you can share some samples to help make testing easier.

Reply:
I can try and grab some samples from testing, but I am not really on the red-team side setting up the o365 attack side and just modifying what another person did with the o365 toolkit they downloaded directly off github repo.

If Microsoft has someone more familiar with the process or someone from DART that can assist, that may be useful.

I think the main issue is; if we search for ALL scopes (as the current rule was written) we are detecting an attack from someone who runs the toolkit "_out-of-the-box._" A smart or advanced actor could simply view this same source code we are editing, since public, or stand up their own Azure account and test and realize that scoping to just a few permissions they need for their attack would not trigger an alert in Sentinel and may go unnoticed?

Reply:
@treyperrone Thank you so much for clarifying. My intention here is to understand what is the minimum scope required to perform this activity. This would help us in performing "has_all" operation. Otherwise, we would end up missing a new of these instances.
I will try to see if we have someone internally who can help us here but if you have the required information, that would be great,
Also, please send the sample logs to demehra@microsoft.com

Reply:
- `has_all` does indeed need all items to fire the alert
- `has_any` may be more appropriate
- we also need to modify [https://github.com/Azure/Azure-Sentinel/blob/cff4ebd4b3efe69331d385d3bf5fa9a88f0c1d35/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L34](https://github.com/Azure/Azure-Sentinel/blob/cff4ebd4b3efe69331d385d3bf5fa9a88f0c1d35/Solutions/Azure%20Active%20Directory/Analytic%20Rules/MaliciousOAuthApp_O365AttackToolkit.yaml#L34) and perhaps fully remove `| where TargetResources has "<some value>"` as it may not serve much of a purpose with searching for all values.

**I still dont know WHAT if all values should be the correct ones here.** With the simple test of only `mail.send` someone can only send mail as that user, but that alone can be damaging if its a malicious attacker sending mail masquerading as an actual user. 

more verbose output of a query test with ONLY `mail.send` 
[query_data_full-output.csv](https://github.com/Azure/Azure-Sentinel/files/10843437/query_data_full-output.csv)

screenshot with `has_any` only (no match)
<img width="1474" alt="Screenshot 2023-02-27 at 2 37 20 PM" src="https://user-images.githubusercontent.com/20403061/221668557-7243c908-4f76-4ebb-93a8-f84a512d4c22.png">


screenshot with `has_any` only (no match)
<img width="1423" alt="Screenshot 2023-02-27 at 2 38 13 PM" src="https://user-images.githubusercontent.com/20403061/221668617-195854b5-1c24-4af5-ae28-7204cfd279f7.png">


screenshot with single dynamic array all values, `has_any` detect, removal of `| where TargetResources has "value"`
<img width="1209" alt="Screenshot 2023-02-27 at 2 56 08 PM" src="https://user-images.githubusercontent.com/20403061/221669589-7e377889-6f1a-4302-ae0b-51bc56040df2.png">

log from this last event above: 
[query_data_single_array.csv](https://github.com/Azure/Azure-Sentinel/files/10843506/query_data_single_array.csv)

Reply:
@devikamehra  see comments above and new updates to the KQL.

Reply:
@treyperrone Just to ensure we are on the same page, you are referring to https://github.com/mdsecactivebreach/o365-attack-toolkit for simulating the attack in this scenario?
If this is true, we should do the following:
1. Remove the TargetResource comparison line as this is a repetition.
2. Move back to the approach you took earlier with two lists. In the attack2 scenario, remove the file read related permissions as these are already inherited with ReadWrite. And now use has_all with respective lists. 

Also, the record that you sent doesn't seems to be the result of the attack documented in the link. If we have has_any operator with such wide scoped grants, the rule will have high number of false positives.

Please let me know in case more information is required from my end.

Reply:
- Yes, we used the attack-toolkit
- the pentester scoped the permissions to only `mail.send` for the example
 
 if we set it back to looking for two `has_all` variants, all we are detecting is an attacker that is not smart enough to work around the sentinel detection we have publicly available on github, so the detection is fair less useful against advanced or persistent attackers

Reply:
@treyperrone Understanding the concern here, we can modify the query in the following way

let detectionTime = 1d;  let joinLookback = 14d;  let threshold = 5;  let o365_attack_regex = "contacts.read|user.read|mail.read|notes.read.all|mailboxsettings.readwrite|Files.ReadWrite.All|mail.send|files.read|files.read.all";  let o365_attack = dynamic(["contacts.read","user.read","mail.read","notes.read.all","mailboxsettings.readwrite","Files.ReadWrite.All","mail.send","files.read","files.read.all"]);  let apps = dynamic(['5523c0ac-586d-491e-bc73-fc18f9ed2c28','3190d40e-82d9-44ee-af78-4ce2aafd4f15','0771c6f9-71c7-46d4-87a9-3b57443ee250','a644e37f-d354-440f-843a-1e1ee7dfc4ef','21fbe0e7-6e83-4fb0-a56f-92a67c64e37a','30966914-6789-4d1c-9e2b-76a81a5e2a54','2a6082d8-9f0a-455c-a26b-7fce81edefb8','adc8d359-f3ef-4e0a-b170-2bcf5a022728']);  AuditLogs  | where TimeGenerated > ago(detectionTime)  | where LoggedByService =~ "Core Directory"  | where Category =~ "ApplicationManagement"  | where OperationName =~ "Consent to application"  | extend AppDisplayName = TargetResources.[0].displayName  | extend AppClientId = tolower(TargetResources.[0].id)  | where AppClientId !in (apps)  | extend ConsentFull = TargetResources[0].modifiedProperties[4].newValue  | parse ConsentFull with * "ConsentType: " GrantConsentType ", Scope: " GrantScope1 ", CreatedDateTime" * "]" *  | where GrantConsentType != "AllPrincipals" // NOTE: we are ignoring if OAuth application was granted to all users via an admin - but admin due diligence should be audited occasionally  | where ConsentFull has_any (o365_attack)  | extend GrantScopeCount = countof(tolower(GrantScope1), o365_attack_regex, 'regex')  | where GrantScopeCount > threshold

If we perform check on has_any, this alert can become noisy and thus would create a huge list of false positive. Thus, adding a threshold here which can prevent this from happening. If you feel the version you suggested would work better in your environment, please tune the threshold parameter to 1 and it should solve the problem.

Let me know if you have any more questions.

Reply:
Sorry for the delay, trying to get more input from other security team members in our organization with more diverse experience and opinions :-)

The threshold suggestion is an excellent one that sets the bar for the average Sentinel user and lets everyone tweak it with ease!

Migrating the external watchlist inline also is good, because the rules are versioned but a change to the external CSV wouldn't be tracked. I would suggest we a minor name tweak to `trusted_apps`, a list format with comments/names for easier end user experience (the CSV has the names for clarity). This list format also puts the closing parenthesis on a newline so easier to comment values out and keep the dynamic array able to pass a NULL value (minor thing, but helpful for newer KQL users).

so in full the potential full playbook could look like below. If these seems good happy to update the patch-2 branch again

```
// let detectionTime = 1d;
// let joinLookback = 14d;
let threshold = 5;
let o365_attack_regex = "contacts.read|user.read|mail.read|notes.read.all|mailboxsettings.readwrite|Files.ReadWrite.All|mail.send|files.read|files.read.all";
let o365_attack = dynamic(["contacts.read", "user.read", "mail.read", "notes.read.all", "mailboxsettings.readwrite", "Files.ReadWrite.All", "mail.send", "files.read", "files.read.all"]);
// The `trusted_apps` are sample App Client IDs. A MATCH will cause the alert to NOT fire; please modify for your environment!
let trusted_apps = dynamic([
'5523c0ac-586d-491e-bc73-fc18f9ed2c28', // DocuSign NA1
'3190d40e-82d9-44ee-af78-4ce2aafd4f15', // Zoom Video Communications
'0771c6f9-71c7-46d4-87a9-3b57443ee250', // Zoom OneDrive
'a644e37f-d354-440f-843a-1e1ee7dfc4ef', // Windows Defender Security Intelligence
'21fbe0e7-6e83-4fb0-a56f-92a67c64e37a', // Amazon Alexa Connect
'30966914-6789-4d1c-9e2b-76a81a5e2a54', // Barracuda Networks
'2a6082d8-9f0a-455c-a26b-7fce81edefb8', // Favro
'adc8d359-f3ef-4e0a-b170-2bcf5a022728' //ClassFlow Prod OneDrive Personal
]);
AuditLogs
| where TimeGenerated > ago(detectionTime)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Consent to application"
| extend AppDisplayName = TargetResources.[0].displayName
| extend AppClientId = tolower(TargetResources.[0].id)
| where AppClientId !in (trusted_apps)
| extend ConsentFull = TargetResources[0].modifiedProperties[4].newValue
| parse ConsentFull with * "ConsentType: " GrantConsentType ", Scope: " GrantScope1 ", CreatedDateTime" * "]" *
| where GrantConsentType != "AllPrincipals" // NOTE: we are ignoring if OAuth application was granted to all users via an admin - but admin due diligence should be audited occasionally
| where ConsentFull has_any (o365_attack)
| extend GrantScopeCount = countof(tolower(GrantScope1), o365_attack_regex, 'regex')
| where GrantScopeCount > threshold
| extend GrantIpAddress = iff(isnotempty(InitiatedBy.user.ipAddress), tostring(InitiatedBy.user.ipAddress), tostring(InitiatedBy.app.ipAddress))
| extend GrantInitiatedBy = iff(isnotempty(InitiatedBy.user.userPrincipalName), tostring(InitiatedBy.user.userPrincipalName), tostring(InitiatedBy.app.displayName))
//
| extend GrantUserAgent = iff(AdditionalDetails[0].key =~ "User-Agent", tostring(AdditionalDetails[0].value), "")
| project TimeGenerated, GrantConsentType, GrantScope1, GrantInitiatedBy, AppDisplayName, GrantIpAddress, GrantUserAgent, AppClientId, OperationName, ConsentFull, CorrelationId
| join kind = leftouter (AuditLogs
| where TimeGenerated > ago(joinLookback)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Add service principal"
| extend AppClientId = tolower(TargetResources[0].id)
| extend AppReplyURLs = iff(TargetResources[0].modifiedProperties[1].newValue has "AddressType", TargetResources[0].modifiedProperties[1].newValue, "")
| distinct AppClientId, tostring(AppReplyURLs)
)
on AppClientId
| join kind = innerunique (AuditLogs
| where TimeGenerated > ago(joinLookback)
| where LoggedByService =~ "Core Directory"
| where Category =~ "ApplicationManagement"
| where OperationName =~ "Add OAuth2PermissionGrant" or OperationName =~ "Add delegated permission grant"
| extend GrantAuthentication = tostring(TargetResources[0].displayName)
| extend GrantOperation = OperationName
| project GrantAuthentication, GrantOperation, CorrelationId
) on CorrelationId
| project TimeGenerated, GrantConsentType, GrantScope1, GrantInitiatedBy, AppDisplayName, AppReplyURLs, GrantIpAddress, GrantUserAgent, AppClientId, GrantAuthentication, OperationName, GrantOperation, CorrelationId, ConsentFull
| extend timestamp = TimeGenerated, AccountCustomEntity = GrantInitiatedBy, IPCustomEntity = GrantIpAddress
```

Reply:
@treyperrone Thank you for working on this.
I will confirm on the trusted app list once again. You are right, the list should be dynamic without much changes to the query. Each approach has its own advantages and disadvantages.
For removing the filters from the join query, this might have performance impact when data is huge. Let's add those in to avoid any query timeouts.

Reply:
@treyperrone 
1. Let's keep the source of trusted app based on the centralized csv for now. We can improve it in future, if required.
2. No, what I am saying is that we should keep the filters as-is in both part of the joins. There is no need to remove any filters from right side of the query. In case you are facing a performance issue, please let us know and we can take adequate steps.

Looking forward to your updated query.